{"cells":[{"cell_type":"markdown","metadata":{"id":"2nYcJkCgH6cJ"},"source":["# Top2Vec"]},{"cell_type":"code","execution_count":1,"metadata":{"executionInfo":{"elapsed":73428,"status":"ok","timestamp":1688760578979,"user":{"displayName":"max riffi","userId":"15293106091454968129"},"user_tz":-60},"id":"NEnbUAwBH6cO"},"outputs":[],"source":["%%capture\n","!pip install top2vec"]},{"cell_type":"code","execution_count":2,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":10054,"status":"ok","timestamp":1688760589008,"user":{"displayName":"max riffi","userId":"15293106091454968129"},"user_tz":-60},"id":"JDvDvcACKHPK","outputId":"bc6b8a4e-0c3f-4ffe-edbe-f3c4368dac18"},"outputs":[{"name":"stdout","output_type":"stream","text":["Mounted at /content/drive\n","/content/drive/MyDrive/yourrepository\n"]}],"source":["import sys\n","if 'google.colab' in sys.modules:\n","  from google.colab import drive\n","  drive.mount('/content/drive')\n","  %cd /content/drive/MyDrive/yourrepository"]},{"cell_type":"code","execution_count":3,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":133411,"status":"ok","timestamp":1688760722403,"user":{"displayName":"max riffi","userId":"15293106091454968129"},"user_tz":-60},"id":"-JrsGz4tH6cU","outputId":"6928fd3b-4e1c-43eb-a7c1-4a2ebd314fa9"},"outputs":[{"name":"stderr","output_type":"stream","text":["[nltk_data] Downloading package punkt to /root/nltk_data...\n","[nltk_data]   Unzipping tokenizers/punkt.zip.\n","[nltk_data] Downloading package stopwords to /root/nltk_data...\n","[nltk_data]   Unzipping corpora/stopwords.zip.\n","[nltk_data] Downloading package wordnet to /root/nltk_data...\n","[nltk_data] Downloading package omw-1.4 to /root/nltk_data...\n"]}],"source":["import os\n","import sys\n","import re\n","import nltk\n","nltk.download('punkt')\n","\n","parent_directory = os.path.abspath('..')\n","sys.path.append(parent_directory)\n","\n","from gensim.test.utils import common_texts\n","from gensim.corpora.dictionary import Dictionary\n","from top2vec import Top2Vec\n","from Preprocess import preprocess\n","from Plot import plot_df\n","\n","# Run the preprocessing\n","df= preprocess('yourrepository/Data/articles.json')\n","\n","# Create a dictionary from the tokenized documents\n","dictionary = Dictionary(df['content'])\n","\n","common_corpus = [','.join(text).replace(',',' ') for text in df['content']]"]},{"cell_type":"markdown","metadata":{"id":"EYe-RA0HH6cX"},"source":["# Hyperparameter Tuning"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":252,"status":"ok","timestamp":1688406952040,"user":{"displayName":"max riffi","userId":"15293106091454968129"},"user_tz":-60},"id":"uI19lFjw5ROb","outputId":"7afa703f-564a-4577-df1a-0ec20eecf30e"},"outputs":[{"name":"stdout","output_type":"stream","text":["min_count: 100\n","topic_merge_delta: 0.1\n","embedding_model: doc2vec\n","chunk_length: 50\n"]}],"source":["import os\n","import sys\n","import re\n","from gensim.test.utils import common_texts\n","from gensim.corpora.dictionary import Dictionary\n","from top2vec import Top2Vec\n","from Preprocess import preprocess\n","from Plot import plot_df\n","\n","# Run the preprocessing\n","df = preprocess('Data/articles.json')\n","\n","# Create a dictionary from the tokenized documents\n","dictionary = Dictionary(df['content'])\n","\n","common_corpus = [','.join(text).replace(',', ' ') for text in df['content']]\n","\n","# Define the hyperparameters to tune\n","min_count = [10, 50, 100]\n","topic_merge_delta = [0.05, 0.1, 0.2]\n","embedding_model = ['doc2vec', 'universal-sentence-encoder', 'distiluse-base-multilingual-cased']\n","chunk_length = [50, 100, 200]\n","\n","best_coherence = float('-inf')\n","best_params = None\n","\n","# Iterate over all combinations of hyperparameters\n","for count in min_count:\n","    for delta in topic_merge_delta:\n","        for model in embedding_model:\n","            for length in chunk_length:\n","                try:\n","                    # Create and train the Top2Vec model\n","                    top2vec_model = Top2Vec(common_corpus, min_count=count, topic_merge_delta=delta,\n","                                            embedding_model=model, chunk_length=length, verbose=False)\n","\n","                    # Evaluate the model and store the results\n","                    coherence = top2vec_model.get_topic_coherence()\n","\n","                    # Check if the current model has better coherence\n","                    if coherence \u003e best_coherence:\n","                        best_coherence = coherence\n","                        best_params = {\n","                            'min_count': count,\n","                            'topic_merge_delta': delta,\n","                            'embedding_model': model,\n","                            'chunk_length': length\n","                        }\n","                except:\n","                    # Handle any errors during model training\n","                    print(\"An error occurred during model training.\")\n","\n","# Print the best parameters\n","print(\"min_count:\", best_params['min_count'])\n","print(\"topic_merge_delta:\", best_params['topic_merge_delta'])\n","print(\"embedding_model:\", best_params['embedding_model'])\n","print(\"chunk_length:\", best_params['chunk_length'])\n"]},{"cell_type":"markdown","metadata":{"id":"c37nXfhWH6ca"},"source":["# Train Model"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"background_save":true,"base_uri":"https://localhost:8080/"},"id":"XnfSNOSuH6cb"},"outputs":[{"name":"stderr","output_type":"stream","text":["2023-07-08 00:24:51,765 - top2vec - INFO - Pre-processing documents for training\n","INFO:top2vec:Pre-processing documents for training\n","2023-07-08 00:25:00,419 - top2vec - INFO - Creating joint document/word embedding\n","INFO:top2vec:Creating joint document/word embedding\n","2023-07-08 00:32:37,175 - top2vec - INFO - Creating lower dimension embedding of documents\n","INFO:top2vec:Creating lower dimension embedding of documents\n","ERROR:root:Internal Python error in the inspect module.\n","Below is the traceback from this internal error.\n","\n","ERROR:root:Internal Python error in the inspect module.\n","Below is the traceback from this internal error.\n","\n","ERROR:root:Internal Python error in the inspect module.\n","Below is the traceback from this internal error.\n","\n"]},{"name":"stdout","output_type":"stream","text":["Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n","    exec(code_obj, self.user_global_ns, self.user_ns)\n","  File \"\u003cipython-input-24-fc469d4ed563\u003e\", line 6, in \u003ccell line: 6\u003e\n","    model = Top2Vec(common_corpus)\n","  File \"/usr/local/lib/python3.10/dist-packages/top2vec/Top2Vec.py\", line 666, in __init__\n","    self.compute_topics(umap_args=umap_args, hdbscan_args=hdbscan_args, topic_merge_delta=topic_merge_delta)\n","  File \"/usr/local/lib/python3.10/dist-packages/top2vec/Top2Vec.py\", line 1256, in compute_topics\n","    umap_model = umap.UMAP(**umap_args).fit(self.document_vectors)\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 2684, in fit\n","    self.embedding_, aux_data = self._fit_embed_data(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 2717, in _fit_embed_data\n","    return simplicial_set_embedding(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 1156, in simplicial_set_embedding\n","    embedding = optimize_layout_euclidean(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/layouts.py\", line 369, in optimize_layout_euclidean\n","    optimize_fn(\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 487, in _compile_for_args\n","    raise e\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 420, in _compile_for_args\n","    return_val = self.compile(tuple(argtypes))\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 965, in compile\n","    cres = self._compiler.compile(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 125, in compile\n","    status, retval = self._compile_cached(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 139, in _compile_cached\n","    retval = self._compile_core(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 152, in _compile_core\n","    cres = compiler.compile_extra(self.targetdescr.typing_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 716, in compile_extra\n","    return pipeline.compile_extra(func)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 452, in compile_extra\n","    return self._compile_bytecode()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 520, in _compile_bytecode\n","    return self._compile_core()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 499, in _compile_core\n","    raise e\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 486, in _compile_core\n","    pm.run(self.state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 368, in run\n","    raise patched_exception\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 356, in run\n","    self._runPass(idx, pass_inst, state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_lock.py\", line 35, in _acquire_compile_lock\n","    return func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 311, in _runPass\n","    mutated |= check(pss.run_pass, internal_state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 273, in check\n","    mangled = func(compiler_state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/untyped_passes.py\", line 234, in run_pass\n","    rewrites.rewrite_registry.apply('before-inference', state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/rewrites/registry.py\", line 67, in apply\n","    matches = rewrite.match(state.func_ir, block, state.typemap,\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/rewrites/static_getitem.py\", line 21, in match\n","    const = func_ir.infer_constant(expr.index)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 1479, in infer_constant\n","    return self._consts.infer_constant(name)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/consts.py\", line 32, in infer_constant\n","    self._cache[name] = (True, self._do_infer(name))\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/consts.py\", line 66, in _do_infer\n","    const = defn.infer_constant()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 598, in infer_constant\n","    raise ConstantInferenceError('%s' % self, loc=self.loc)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/errors.py\", line 671, in __init__\n","    super(ConstantInferenceError, self).__init__(value, loc=loc)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/errors.py\", line 518, in __init__\n","    new_msg = \"%s\\n%s\\n\" % (msg, loc.strformat())\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 113, in strformat\n","    lines = self.get_lines()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 93, in get_lines\n","    self.lines = linecache.getlines(self._get_path())\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 102, in _get_path\n","    path = os.path.relpath(self.filename)\n","  File \"/usr/lib/python3.10/posixpath.py\", line 488, in relpath\n","    start_list = [x for x in abspath(start).split(sep) if x]\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'OSError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1101, in get_records\n","    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 248, in wrapped\n","    return f(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n","    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n","  File \"/usr/lib/python3.10/inspect.py\", line 1662, in getinnerframes\n","    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n","  File \"/usr/lib/python3.10/inspect.py\", line 1620, in getframeinfo\n","    filename = getsourcefile(frame) or getfile(frame)\n","  File \"/usr/lib/python3.10/inspect.py\", line 829, in getsourcefile\n","    module = getmodule(object, filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 861, in getmodule\n","    file = getabsfile(object, _filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 845, in getabsfile\n","    return os.path.normcase(os.path.abspath(_filename))\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n","    exec(code_obj, self.user_global_ns, self.user_ns)\n","  File \"\u003cipython-input-24-fc469d4ed563\u003e\", line 6, in \u003ccell line: 6\u003e\n","    model = Top2Vec(common_corpus)\n","  File \"/usr/local/lib/python3.10/dist-packages/top2vec/Top2Vec.py\", line 666, in __init__\n","    self.compute_topics(umap_args=umap_args, hdbscan_args=hdbscan_args, topic_merge_delta=topic_merge_delta)\n","  File \"/usr/local/lib/python3.10/dist-packages/top2vec/Top2Vec.py\", line 1256, in compute_topics\n","    umap_model = umap.UMAP(**umap_args).fit(self.document_vectors)\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 2684, in fit\n","    self.embedding_, aux_data = self._fit_embed_data(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 2717, in _fit_embed_data\n","    return simplicial_set_embedding(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 1156, in simplicial_set_embedding\n","    embedding = optimize_layout_euclidean(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/layouts.py\", line 369, in optimize_layout_euclidean\n","    optimize_fn(\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 487, in _compile_for_args\n","    raise e\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 420, in _compile_for_args\n","    return_val = self.compile(tuple(argtypes))\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 965, in compile\n","    cres = self._compiler.compile(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 125, in compile\n","    status, retval = self._compile_cached(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 139, in _compile_cached\n","    retval = self._compile_core(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 152, in _compile_core\n","    cres = compiler.compile_extra(self.targetdescr.typing_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 716, in compile_extra\n","    return pipeline.compile_extra(func)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 452, in compile_extra\n","    return self._compile_bytecode()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 520, in _compile_bytecode\n","    return self._compile_core()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 499, in _compile_core\n","    raise e\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 486, in _compile_core\n","    pm.run(self.state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 368, in run\n","    raise patched_exception\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 356, in run\n","    self._runPass(idx, pass_inst, state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_lock.py\", line 35, in _acquire_compile_lock\n","    return func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 311, in _runPass\n","    mutated |= check(pss.run_pass, internal_state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 273, in check\n","    mangled = func(compiler_state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/untyped_passes.py\", line 234, in run_pass\n","    rewrites.rewrite_registry.apply('before-inference', state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/rewrites/registry.py\", line 67, in apply\n","    matches = rewrite.match(state.func_ir, block, state.typemap,\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/rewrites/static_getitem.py\", line 21, in match\n","    const = func_ir.infer_constant(expr.index)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 1479, in infer_constant\n","    return self._consts.infer_constant(name)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/consts.py\", line 32, in infer_constant\n","    self._cache[name] = (True, self._do_infer(name))\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/consts.py\", line 66, in _do_infer\n","    const = defn.infer_constant()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 598, in infer_constant\n","    raise ConstantInferenceError('%s' % self, loc=self.loc)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/errors.py\", line 671, in __init__\n","    super(ConstantInferenceError, self).__init__(value, loc=loc)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/errors.py\", line 518, in __init__\n","    new_msg = \"%s\\n%s\\n\" % (msg, loc.strformat())\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 113, in strformat\n","    lines = self.get_lines()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 93, in get_lines\n","    self.lines = linecache.getlines(self._get_path())\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 102, in _get_path\n","    path = os.path.relpath(self.filename)\n","  File \"/usr/lib/python3.10/posixpath.py\", line 488, in relpath\n","    start_list = [x for x in abspath(start).split(sep) if x]\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'OSError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n","    if (await self.run_code(code, result,  async_=asy)):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3575, in run_code\n","    self.showtraceback(running_compiled_code=True)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2101, in showtraceback\n","    stb = self.InteractiveTB.structured_traceback(etype,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n","    return FormattedTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n","    return VerboseTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1124, in structured_traceback\n","    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n","    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n","    return len(records), 0\n","TypeError: object of type 'NoneType' has no len()\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1101, in get_records\n","    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 248, in wrapped\n","    return f(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n","    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n","  File \"/usr/lib/python3.10/inspect.py\", line 1662, in getinnerframes\n","    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n","  File \"/usr/lib/python3.10/inspect.py\", line 1620, in getframeinfo\n","    filename = getsourcefile(frame) or getfile(frame)\n","  File \"/usr/lib/python3.10/inspect.py\", line 829, in getsourcefile\n","    module = getmodule(object, filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 861, in getmodule\n","    file = getabsfile(object, _filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 845, in getabsfile\n","    return os.path.normcase(os.path.abspath(_filename))\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n","    exec(code_obj, self.user_global_ns, self.user_ns)\n","  File \"\u003cipython-input-24-fc469d4ed563\u003e\", line 6, in \u003ccell line: 6\u003e\n","    model = Top2Vec(common_corpus)\n","  File \"/usr/local/lib/python3.10/dist-packages/top2vec/Top2Vec.py\", line 666, in __init__\n","    self.compute_topics(umap_args=umap_args, hdbscan_args=hdbscan_args, topic_merge_delta=topic_merge_delta)\n","  File \"/usr/local/lib/python3.10/dist-packages/top2vec/Top2Vec.py\", line 1256, in compute_topics\n","    umap_model = umap.UMAP(**umap_args).fit(self.document_vectors)\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 2684, in fit\n","    self.embedding_, aux_data = self._fit_embed_data(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 2717, in _fit_embed_data\n","    return simplicial_set_embedding(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/umap_.py\", line 1156, in simplicial_set_embedding\n","    embedding = optimize_layout_euclidean(\n","  File \"/usr/local/lib/python3.10/dist-packages/umap/layouts.py\", line 369, in optimize_layout_euclidean\n","    optimize_fn(\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 487, in _compile_for_args\n","    raise e\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 420, in _compile_for_args\n","    return_val = self.compile(tuple(argtypes))\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 965, in compile\n","    cres = self._compiler.compile(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 125, in compile\n","    status, retval = self._compile_cached(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 139, in _compile_cached\n","    retval = self._compile_core(args, return_type)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/dispatcher.py\", line 152, in _compile_core\n","    cres = compiler.compile_extra(self.targetdescr.typing_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 716, in compile_extra\n","    return pipeline.compile_extra(func)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 452, in compile_extra\n","    return self._compile_bytecode()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 520, in _compile_bytecode\n","    return self._compile_core()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 499, in _compile_core\n","    raise e\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler.py\", line 486, in _compile_core\n","    pm.run(self.state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 368, in run\n","    raise patched_exception\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 356, in run\n","    self._runPass(idx, pass_inst, state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_lock.py\", line 35, in _acquire_compile_lock\n","    return func(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 311, in _runPass\n","    mutated |= check(pss.run_pass, internal_state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/compiler_machinery.py\", line 273, in check\n","    mangled = func(compiler_state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/untyped_passes.py\", line 234, in run_pass\n","    rewrites.rewrite_registry.apply('before-inference', state)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/rewrites/registry.py\", line 67, in apply\n","    matches = rewrite.match(state.func_ir, block, state.typemap,\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/rewrites/static_getitem.py\", line 21, in match\n","    const = func_ir.infer_constant(expr.index)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 1479, in infer_constant\n","    return self._consts.infer_constant(name)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/consts.py\", line 32, in infer_constant\n","    self._cache[name] = (True, self._do_infer(name))\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/consts.py\", line 66, in _do_infer\n","    const = defn.infer_constant()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 598, in infer_constant\n","    raise ConstantInferenceError('%s' % self, loc=self.loc)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/errors.py\", line 671, in __init__\n","    super(ConstantInferenceError, self).__init__(value, loc=loc)\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/errors.py\", line 518, in __init__\n","    new_msg = \"%s\\n%s\\n\" % (msg, loc.strformat())\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 113, in strformat\n","    lines = self.get_lines()\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 93, in get_lines\n","    self.lines = linecache.getlines(self._get_path())\n","  File \"/usr/local/lib/python3.10/dist-packages/numba/core/ir.py\", line 102, in _get_path\n","    path = os.path.relpath(self.filename)\n","  File \"/usr/lib/python3.10/posixpath.py\", line 488, in relpath\n","    start_list = [x for x in abspath(start).split(sep) if x]\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'OSError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n","    if (await self.run_code(code, result,  async_=asy)):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3575, in run_code\n","    self.showtraceback(running_compiled_code=True)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2101, in showtraceback\n","    stb = self.InteractiveTB.structured_traceback(etype,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n","    return FormattedTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n","    return VerboseTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1124, in structured_traceback\n","    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n","    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n","    return len(records), 0\n","TypeError: object of type 'NoneType' has no len()\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3030, in _run_cell\n","    return runner(coro)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n","    coro.send(None)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3257, in run_cell_async\n","    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3492, in run_ast_nodes\n","    self.showtraceback()\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2101, in showtraceback\n","    stb = self.InteractiveTB.structured_traceback(etype,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n","    return FormattedTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n","    return VerboseTB.structured_traceback(\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1142, in structured_traceback\n","    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n","    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n","    return len(records), 0\n","TypeError: object of type 'NoneType' has no len()\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2099, in showtraceback\n","    stb = value._render_traceback_()\n","AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n","\n","During handling of the above exception, another exception occurred:\n","\n","Traceback (most recent call last):\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 1101, in get_records\n","    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 248, in wrapped\n","    return f(*args, **kwargs)\n","  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n","    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n","  File \"/usr/lib/python3.10/inspect.py\", line 1662, in getinnerframes\n","    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n","  File \"/usr/lib/python3.10/inspect.py\", line 1620, in getframeinfo\n","    filename = getsourcefile(frame) or getfile(frame)\n","  File \"/usr/lib/python3.10/inspect.py\", line 829, in getsourcefile\n","    module = getmodule(object, filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 861, in getmodule\n","    file = getabsfile(object, _filename)\n","  File \"/usr/lib/python3.10/inspect.py\", line 845, in getabsfile\n","    return os.path.normcase(os.path.abspath(_filename))\n","  File \"/usr/lib/python3.10/posixpath.py\", line 384, in abspath\n","    cwd = os.getcwd()\n","OSError: [Errno 107] Transport endpoint is not connected\n"]}],"source":["min_count = 100\n","topic_merge_delta = 0.1\n","embedding_model = 'doc2vec'\n","chunk_length = 50\n","\n","model = Top2Vec(common_corpus)\n","\n","model.save(\"Top2vec_Model\")"]},{"cell_type":"markdown","metadata":{"id":"w_CipY-VH6cb"},"source":["# Coherence Score"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Rw89YsOTH6cc"},"outputs":[],"source":["topic_words = model.get_topics(num_topics=Num_topics, reduced=False)[0].tolist()\n","\n","from Plot import print_coherence\n","dic = dictionary\n","print_coherence(dic, topics_words[0], df['content'])"]},{"cell_type":"markdown","metadata":{"id":"V98qx-OuH6cc"},"source":["# Dynamic Topic modelling"]},{"cell_type":"markdown","metadata":{"id":"A7JSWB30H6cc"},"source":["Some extra preprocessing was required due to the function get_documents_topics from the Top2vec API.\n","\n","'get_documents_topics'\n","\n","Returns:\n","topic_nums (array of int, shape(len(doc_ids), num_topics)) – The topic number(s) of the document corresponding to each doc_id.\n","\n","topic_score (array of float, shape(len(doc_ids), num_topics)) – Semantic similarity of document to topic(s). The cosine similarity of the document and topic vector.\n","\n","topics_words (array of shape(len(doc_ids), num_topics, 50)) – For each topic the top 50 words are returned, in order of semantic similarity to topic.\n","\n"]},{"cell_type":"code","execution_count":10,"metadata":{"executionInfo":{"elapsed":13920,"status":"ok","timestamp":1688761640009,"user":{"displayName":"max riffi","userId":"15293106091454968129"},"user_tz":-60},"id":"N2b3zeRoH6cc"},"outputs":[],"source":["import numpy as np\n","\n","df_distribution = df\n","\n","\n","num_docs = len(df_distribution)  # Get the number of documents in the DataFrame\n","\n","# Reset the index and create a new column for document IDs\n","df_distribution = df_distribution.reset_index()\n","df_distribution.rename(columns={'index': 'doc_id'}, inplace=True)\n","\n","# Call the get_documents_topics function\n","topic_distribution = model.get_documents_topics([i for i in range(num_docs)], num_topics=model.get_num_topics())\n","\n","topic_nums, topic_scores, topics_words, word_distribution = topic_distribution[0], topic_distribution[1], topic_distribution[2], topic_distribution[3]\n","\n","# Insert topic columns in the data\n","for i in range(num_docs):\n","    # Retrieve the topic numbers and scores for the current document\n","    doc_topic_nums = topic_nums[i].tolist()\n","    doc_topic_scores = topic_scores[i].tolist()\n","    topics = topics_words[i].tolist()\n","    topic_hierarchy = word_distribution[i].tolist()\n","\n","    # Insert a column for each topic in the data\n","    for j, topic_num in enumerate(doc_topic_nums):\n","        topic_score = doc_topic_scores[j]\n","        topic_name =  topics[j]\n","        word_distri = topic_hierarchy[j]\n","\n","\n","        a = sorted(word_distri, reverse = True)[:4]\n","        Topic_tile = [topic_name[word_distri.index(i)] for i in a][:4]\n","\n","        # Ensure the column name is a string\n","        column_name = f'{Topic_tile}'\n","\n","        # Check if the column already exists\n","        if column_name not in df_distribution.columns:\n","            df_distribution[column_name] = np.nan\n","\n","        # Convert the topic score to numpy.float32\n","        topic_score = np.float32(topic_score)\n","\n","        # Insert the topic number and score into your data\n","        df_distribution.at[i, column_name] = topic_score"]},{"cell_type":"code","execution_count":23,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000,"output_embedded_package_id":"17IsQLZjsCWRjJdG5Ml7GKFEYg_et2tYJ"},"executionInfo":{"elapsed":58602,"status":"ok","timestamp":1688771447868,"user":{"displayName":"max riffi","userId":"15293106091454968129"},"user_tz":-60},"id":"kDrp0ab3H6cd","outputId":"f5e97423-3938-4f8a-9302-3d3841a165f8"},"outputs":[{"data":{"text/plain":"Output hidden; open in https://colab.research.google.com to view."},"metadata":{},"output_type":"display_data"}],"source":["  %cd /content/drive/MyDrive/yourrepository/Dynamic_Top2Vec\n","  topic_means = df_distribution.groupby('year-month').mean()\n","\n","plot_df(topic_means,'Top2vec', add_error_bars =False)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jtlHDcX_pqiD"},"outputs":[],"source":[]}],"metadata":{"colab":{"name":"","version":""},"kernelspec":{"display_name":"Python [conda env:Python3] *","language":"python","name":"conda-env-Python3-py"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.15"}},"nbformat":4,"nbformat_minor":0}